### PR TITLE
add addField methode to CRM_Core_Form CRM-16130

### PR DESCRIPTION
**Pull request by mallezie**
https://issues.civicrm.org/jira/browse/CRM-16130

First version which supports some fields, not yet all, but more of addSelect. Has first conversions. 
I think we should start converting forms from here, and add functioniality to the method when needed. 
2 possible goals should be to use this as much as possible, and also use it for customfields.

### CODE REVIEW

**Comment by colemanw** [comment]
This is exciting progress :)

File: CRM/Core/Form.php [hunk 1]
```diff
@@ -1109,6 +1109,142 @@ public function addSelect($name, $props = array(), $required = FALSE) {
...
+    $info = civicrm_api3($props['entity'], 'getfields');
+    foreach ($info['values'] as $uniqueName => $fieldSpec) {
+      if (
+          $uniqueName === $props['field'] ||
+          CRM_Utils_Array::value('name', $fieldSpec) === $props['field'] ||
+          in_array($props['field'], CRM_Utils_Array::value('api.aliases', $fieldSpec, array()))
+      ) {
+        break;
+      }
+    }
```
Comment (colemanw) [inline CRM/Core/Form.php]: I think resolving field properties deserves a helper function to reduce the size of this function and be more reusable. I'll work on that...

File: CRM/Core/Form.php [hunk 2]
```diff
@@ -1109,6 +1109,142 @@ public function addSelect($name, $props = array(), $required = FALSE) {
...
+
+    $label = isset($props['label']) ? $props['label'] : $fieldSpec['title'];
+
+    if (!array_key_exists('placeholder', $props)) {
+      $props['placeholder'] = $required ? ts('- select -') : CRM_Utils_Array::value('context', $props) == 'search' ? ts('- any -') : ts('- none -');
```
Comment (colemanw) [inline CRM/Core/Form.php]: Seems like this placeholder is only appropriate for select elements.

File: CRM/Core/Form.php [hunk 3]
```diff
@@ -1109,6 +1109,142 @@ public function addSelect($name, $props = array(), $required = FALSE) {
...
+
+      case 'TextArea':
+        //Set default rows and cols
+        $props['rows'] = isset($props['rows']) ? $props['rows'] : 4;
+        $props['cols'] = isset($props['cols']) ? $props['cols'] : 60;
+        $this->add('textarea', $name, $label, NULL, $required, $props);
+        break;
+
+      case 'Select Date':
+        //TODO: Add date formats
```
Comment (colemanw) [inline CRM/Core/Form.php]: Yea another todo is to remove the dependency of manually including the datepicker js in the template.
I've got a separate issue open for that somewhere...

**Comment by mallezie** [comment]
Thanks @colemanw for the review. I've fixed / added TODO's for your remarks. What should be the next steps? Finish more of the method here? Or merge this, and start converting forms, adding more type's when we go? 
I'm not sure how far i should go in this PR? 
Do you want me to file followup issues for this? For example for the javascript adding / refactoring the helper function for getting the field properties? Another one would be to make the separations to add fields on search forms better? The ```CRM_Utils_Array::value('context', $props) == 'search'``` clauses.

**Comment by colemanw** [comment]
Good questions.

I think we can keep the existing issue, but agree that this PR needs a stopping point where we merge it and then start a new one. I vote for keeping it open for a few more commits before merging.

It would make collaboration a little easier if you give me write access to your repo (as long as you are familiar enough with git to be able to rebase your local commit if you get a non-ff push rejection).

I'll work on the helper separately and we can incorporate that in the next PR.

I agree that a good next step would be to have the 'context' code a bit cleaner. How about making it a property of the form class similar to how you did the defaultEntity?

**Comment by colemanw** [comment]
Since one form is only going to have one context, probably ok to have it be a simple class property and not a getter function.

**Comment by mallezie** [comment]
Thanks for the answers @colemanw 
I've given you access to my repo. Could you reference the PR for the helper here, then we can use it here before merging. 
Perhaps we should continue this PR untill we have one form fully converted to only use addField. It could be the individual form where i already started?

**Comment by mallezie** [comment]
Perhaps we do should go for the getter function, this makes it possibly to override a form and adding a dynamically getContext function. Suppose i want to make a form for two contexts, then i can do 

´´´class myform extends core form{
function getContext(){
  $context = parent::getContext();
  if(url_of_this_form() = 'blablasearch'){
    return 'search;
  }
  else{
    return $context
  }
}
´´´
Something i can do now with the entity, but wouldn't be able to do with the class property. 
I'm not sure if it's a good thing to do, but it could be possible.

**Comment by colemanw** [comment]
Ok you're right, that would be more flexible.

**Comment by colemanw** [comment]
I'm a little skeptical of refactoring all of the add contact form. Might want to pick an easier target for the first round.

**Comment by mallezie** [comment]
I was thinking of the individual only (which is only part of the contact form), but that's indeed quite a crucial one. Any suggestion for a form? We should pick one with at least some different form element type's.

**Comment by colemanw** [comment]
Ok, yea Individual sub-form isn't so bad, and presents a few nice challenges.

1. Some fields in xml/schema/Contact/Contact.xml do not yet have an <html> tag. Or is incomplete, or wrong. The tag is fairly new (added at last year's code sprint) and we haven't started using it for much of anything... until now! Note that when changing a schema xml file you must run `cd xml && php GenCode.php` to have the changes propagate into the DAO files.

2. The `employer_id` should probably be changed from "Autocomplete-Select" to "EntityRef" and we should probably change this for all fields with that html type declared (in retrospect it's not a very accurate label. I think we chose it to be consistent with custom fields, but I vote to change it anyway). The $props being passed to addEntityRef in the form should probably be added as tags inside the <html> tag. The $entity can be inferred from the FKApiName attribute (which ought to come through in the api.getfields data), same as how the api explorer does it: https://github.com/civicrm/civicrm-core/blob/master/templates/CRM/Admin/Page/APIExplorer.js#L362

Is that all making sense? I'm happy to expand on any explanations if I'm not providing enough info.

**Comment by mallezie** [comment]
Thanks i'll give it a further try.

**Comment by mallezie** [comment]
I added some further conversions of form fields, and did some cleanup. 
I'm not sure to use about which method to use to add the form element in the addField method. Should i try ot use the quickform method addElement, or should it be good to use the own CRM_Core_Form methods (add / addDate / ...), perhaps with the goal to make them internal (protected) functions?

File: CRM/Core/Form.php [hunk 4]
```diff
@@ -1109,6 +1119,141 @@ public function addSelect($name, $props = array(), $required = FALSE) {
...
+      // Add data for popup link.
+      if ($props['context'] != 'search' && $widget == 'Select' && CRM_Core_Permission::check('administer CiviCRM')) {
+        $props['data-option-edit-path'] = array_key_exists('option_url', $props) ? $props['option_url'] : $props['data-option-edit-path'] = CRM_Core_PseudoConstant::getOptionEditUrl($fieldSpec);
+        $props['data-api-entity'] = $props['entity'];
+        $props['data-api-field'] = $props['field'];
+      }
+    }
+
+    CRM_Utils_Array::remove($props, 'entity', 'field', 'context', 'label');
+    switch ($widget) {
```
Comment (totten) [inline CRM/Core/Form.php]: A sense of deja vu:
 * https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/BAO/CustomField.php#L828
 * https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/BAO/UFGroup.php#L1837

IMHO, these switch/if-else structures are significant to understanding Civi-Quckform's history/mistakes. The switch() statement is quite understandable as an intermediate step, but when it lingers in the codebase it becomes something scary that many developers avoid (prob due to a mix of TLDR-itis and conservatism... but maybe I'm projecting...). Anyway, it's pretty important that form components be represented in decoupled, bite-sized chunks (like QuickForm elements or jQuery plugins or Angular directives, ad nauseum).

With that said, maybe the simple act of moving the switch from CRM_Core_BAO_CustomGroup to CRM_Core_Form would make a major change in perspective...

**Comment by mallezie** [comment]
@totten it's true its 'copied' from CRM_Core_BAO_CustomGroup. Goal is indeed to remove it from there, or to replace with a wrapper for addField. But that would be to much for one PR, since the structure of CustomFields isn't really compatible with the structure of basic fields.

**Comment by mallezie** [comment]
I converted all fields of the individual form to use addField. The addField method doesn't support all fields yet, but that's perhaps better to use with an other form conversion, and go on till everything is converted. 
any more objections to make this merge-ready? 
I also left the entityRef / Autocomplete select change as is for now.

**Comment by colemanw** [comment]
I want to keep the momentum going so I'll merge this. Looking forward to the next installment :)

**Comment by mallezie** [comment]
Super! this is exciting! i'll file new issues and PR for the next forms, any prefered forms to do first? Else i'll go with the other contact forms first. 
Is there some sort of policy to create [Meta] issues to keep an overview on which forms still needs to be done? Else i'll create an issue with all forms as sub-issues, where the sub-issues can then be taken one by one? That way i can create an overview what needs to be done.

**Comment by colemanw** [comment]
There doesn't have to be a 1-1 ratio between issues and PRs, it's not unusual for 1 issue in jira to have several PRs.
Jira does have a way of creating "sub-tasks" which we use for large projects. This is probably a good case for using them, but I wouldn't go crazy. Sometimes a bulleted list in the issue description works just as well. Or sometimes a quick google spreadsheet can be handy for tracking todos. I'd say do whatever feels genuinely helpful.

**Pull request merged**